Passenger Enterprise support, fix `passenger-config` not being run with 'cd #{release_path}'

FooBarWidget opened the pull request:
This pull request fixes two issues.
- Fixes parsing Passenger Enterprise version output.
- Fix `passenger-config` not being run with 'cd #{release_path}' if passenger_in_gemfile is true.
  
  The old method tries to obtain a command string using SSHKit::Command.new. That doesn't work so well: the `within(release_path)` rule isn't applied to the command string. As a result, Capistrano would try to run "bundle exec passenger-config" without prepending "cd #{release_path}". This new code fixes that.

Can I request a new release ASAP? I found this bug while testing the new Passenger documentation (https://www.phusionpassenger.com/library/walkthroughs/deploy/ruby/ownserver/standalone/oss/install_language_runtime.html and https://www.phusionpassenger.com/library/advanced_production/ruby_capistrano.html). I'd like to be able to add a version specifier in my example code so that users are guaranteed to get a working example.

FooBarWidget committed:
Support parsing Passenger Enterprise version output

FooBarWidget committed:
Fix `passenger-config` not being run with 'cd #{release_path}' when passenger_in_gemfile is true

The old method preprocesses the command with SSHKit::Command.new
in order to allow the 'passenger-config' command to be transformed
via the SSHKit command map, even when the entire command is prepended
with 'sudo'. However, that approach broke cases in which
passenger_restart_with_sudo is false, because a string (instead of a
command symbol) is passed to `execute`. As a result, `execute` does
not prepend `cd #{release_path}` to the final command. This would
normally not cause problems, but would cause problems when passenger_in_gemfile
is true because 'bundle exec passenger-config' tries to look for a Gemfile
in the current working directory.

betesh commented:
1) This conflicts with #25.  Is it really necessary to use `\A` in the regex?

2) This also appears to undo 7127d493a8cd.  That was a cherry-pick of the 3rd commit in #16, so please see the discussion at https://github.com/capistrano/passenger/pull/16#issuecomment-107947207 and make sure it doesn't undo support for using sudo with chruby.  One possible approach: if `fetch(:passenger_restart_with_sudo) == true`, then for chruby, after capistrano-chruby runs its hooks, we do:

```
SSHKit.config.command_map.prefix[:'passenger-config'].unshift(:sudo)
```

FooBarWidget commented:
Oh whoops. I kind of copy-pasted my changes from the latest release version of capistrano-passenger. I didn't check against git master, so I overwrote any patches already in there. Let me check again.

FooBarWidget commented:
All right, fixed both problems. Turns out 7127d49 introduced a bug, which I have fixed now. Please see the commit message for a detailed analysis.

I have confirmed that my code works with both RVM and chruby, with and without sudo.

betesh commented:
Can you avoid the duplication of `fetch(:passenger_restart_command).split(" ").collect(&:to_sym)`?

FooBarWidget commented:
Done.

betesh commented:
I did some refactoring of my own and pushed it as pr27 branch. (Diff: https://github.com/capistrano/passenger/compare/master...pr27)

> I have confirmed that my code works with both RVM and chruby, with and without sudo.

@FooBarWidget Do you have a ready-to-go test environment where you can repeat that quickly with the pr27 branch?

FooBarWidget commented:
Yes, everything works fine with the pr27 branch.
